README.md
# Batteries Not Included

Batteries Not Included is a different take on electric vehicle mods.  Rather than using equipment grids and/or wireless
charging, it operates on the premise of quickly-swappable batteries that can be recharged at a dedicated charging 
station.

This internally uses a similar mechanism to that of uranium fuel cells and nuclear reactors.  You insert battery packs
 into electric trains and -- when they're drained -- they produce empty battery packs which can then be removed and
 brought to a charging station.
 
The end goal for this mod is to provide alternatives to solid fuel and rocket fuel in trains which take advantage of
capabilities commonly found in electric engines (such as rapid acceleration).  These benefits are intended to be 
balanced out by both production costs and the logistics overhead of setting up a charging infrastructure.

## HELP WANTED

All of the art in this mod is currently placeholder graphics and/or my terrible paint.net skills.  Contributions to 
 making it less ugly are welcome.
  
Discussions about balance are also welcome!  Find me on Discord or the forums.
  
## Battery Packs

To get started, you'll need to make at least one battery pack (probably much more).  Begin by researching
"Heavy-Duty Battery Packs", which requires research in batteries, electric energy accumulators and electric energy
 distribution 2.  
 
 After that, you can create drained battery packs.  Wait, drained?  Yes.  Before you can actually use 
   your shiny new battery pack, you'll need to charge it.  This requires a Charging Station (which currently 
   looks suspiciously like a blue accumulator -- did you see HELP WANTED?) -- insert a drained battery pack into a charging station, wait for awhile, and a 
   charged battery pack comes out.
   
**Note**: Other than drain, the power usage for charge stations (both types!) actually matches up with how power is
going into a battery being charged. 

## Electric Locomotives

You'll need Electric Locomotives to be able to use Battery Packs.  Note that they only hold 20 battery packs (batteries
have far worse energy density than fossil fuels).  You can insert battery packs into locomotives just like fuel in a
regular locomotive, and remove empty battery packs as well.  Inserters won't remove full battery packs.

## High-Performance Battery Packs

These fancy gold-label batteries hold more power and offer substantially increased vehicle performance.  Of course, 
this means more charging time...

## Rapid Charging Stations

Charge batteries, faster.

## Known Issues

* Placeholder graphics are terrible.  Did you read HELP WANTED?
  
* Balance needs work.  Did you read HELP WANTED?!

## Unknown Issues
   
Found a bug?  Please visit the [Issues page](https://github.com/dewiniaid/BatteriesNotIncluded/issues) to see if it has 
already been reported, and report it if not.

## Changelog

### 0.2.2 (2020-02-01)
* **Update for Factorio 0.18**

### 0.2.1 (2019-03-31)
* Update research requirements for changes made in Factorio 0.17.23
* Fix crash when adding Batteries Not Included to an existing save that had QuickItemSwap enabled.

### 0.2.0 (2019-02-26)
* Updated for Factorio 0.17

### 0.1.4 (2018-10-08)
* Added German translations.  Thanks to MrBesen and SuperSandro2000 for contributing.
* Partially-depleted batteries are no longer completely lost when mining a vehicle and are returned in the form of
  a fully-depleted battery instead.
* Increased amount of cables, circuits and advanced circuits required for making charging stations.
* Support QuickItemSwap natively. (QIS already supports Batteries Not Included, so this change should be invisible.)

### 0.1.3 (2017-12-23)
* Rapid Charging Station rebalanced: Now requires 15 advanced processing units (reduced from 20), has 250% speed (increased from 200), and 80% efficiency (reduced from 100%).
* Charging stations no longer produce drain.

### 0.1.2 (2017-12-14)
* Fixed the Electric Locomotive being available without research.  (It's useless without the research for battery packs anyways).
* Fixed Heavy Duty Battery Packs to require Electric Energy Distribution 1 instead of Electric Energy Distribution 2.  It no longer is a green science recipe depending on a blue science recipe.
* Renamed "Heavy Duty Battery Packs" technology to "Battery Packs".  Heavy-Duty Battery Packs may appear in another incarnation in the future.
* Removed research dependency on batteries, since it is implied by also depending on Accumulators.
* Support Factorio 0.16 native changelog features.

### 0.1.1 (2017-12-13)
* Update for Factorio 0.16
* Temporarily remove support for QuickItemSwap to work around [a bug in the current Factorio experimental](https://forums.factorio.com/viewtopic.php?f=182&t=54567&p=321491).  To counteract this, support is now provided in QuickItemSwap itself. 

### 0.1.0 (2017-12-11)
* First release.





Blueprint Extensions 0.4.0 released, with new auto-landfill features

I don't normally toot my own horn as a mod author here, but every now and then I feel like I should share something.  So, without further ado:

It's no joke, I've just released version 0.4.0 of Blueprint Extensions for Factorio 0.17.23+.  **The main new feature in this version is the ability to automatically add landfill to blueprints**, along with some revamped GUI functionality.

For those new to the mod, Blueprint Extensions has a number of other useful quality-of-life tools related to blueprints:

 * **Mirror**: Mirror most blueprints.  (Rails may not mirror cleanly).  This correctly adjusts splitter priorities.  If [Gah!DarnItWater!](https://mods.factorio.com/mod/GDIW) is installed, refinery and chemical plant recipes will be mirrored as well.
 
 * **Snap**: Snap a blueprint so a specific corner or edge of it is under your mouse rather than the middle.  Handy for lining up those large nuclear power plants and tileable solar blueprints you might have.
 
 * **Update**: Quickly create a brand new blueprint using the same name and icons as an existing one, optionally with an auto-incrementing version number.
 
 * **Wireswap**: Swap between red and green wires in your blueprint.
 
 * **Rotate**: Rotate an actual blueprint (rather than just your hand).  Useful if you have a family of related blueprints and want them all facing the same direction in the library.
 
You can find the new version of Blueprint Extensions [on the mod portal](https://mods.factorio.com/mod/BlueprintExtensions), or in game as usual.

Also feel free to check out [Rail Tools](https://mods.factorio.com/mod/RailTools) for rail signal placement and error checking, [GhostBuster](https://mods.factorio.com/mod/GhostBuster) to prevent mods from tearing up belts only to put the same belt back down, or [the rest of my quality-of-life mods](https://mods.factorio.com/user/dewiniaid).


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.2.2
Date: 01. 02. 2020
  Changes:
    - Update for Factorio 0.18

---------------------------------------------------------------------------------------------------
Version: 0.2.1
Date: 31. 03. 2019
  Bugfixes:
    - Update research requirements for changes made in Factorio 0.17.23
    - Fix crash when adding Batteries Not Included to an existing save that had QuickItemSwap enabled.

---------------------------------------------------------------------------------------------------
Version: 0.2.0
Date: 26. 02. 2019
  Changes:
    - Update for Factorio 0.17

---------------------------------------------------------------------------------------------------
Version: 0.1.4
Date: 08. 10. 2018
  Changes:
    - Added German translations.  Thanks to MrBesen and SuperSandro2000 for contributing.
    - Partially-depleted batteries are no longer completely lost when mining a vehicle and are returned in the form of
      a fully-depleted battery instead.
    - Increased amount of cables, circuits and advanced circuits required for making charging stations.
    - Support QuickItemSwap natively. (QIS already supports Batteries Not Included, so this change should be invisible.)

---------------------------------------------------------------------------------------------------
Version: 0.1.3
Date: 14. 23. 2017
  Changes:
    - Rapid Charging Station rebalanced: Now requires 15 advanced processing units (reduced from 20), has 250% speed (increased from 200), and 80% efficiency (reduced from 100%).
    - Charging stations no longer produce drain.

---------------------------------------------------------------------------------------------------
Version: 0.1.2
Date: 14. 12. 2017
  Bugfixes:
    - Fixed Electric Locomotive being available without being researched.
    - Fixed Heavy Duty Battery Packs to require Electric Energy Distribution 1 instead of Electric Energy Distribution 2.  It no longer is a green science recipe depending on a blue science recipe.
  Changes:
    - Renamed "Heavy Duty Battery Packs" technology to "Battery Packs".  Heavy-Duty Battery Packs may appear in another incarnation in the future.
    - Removed research dependency on batteries, since it is implied by also depending on Accumulators.
    - Support Factorio 0.16 native changelog features.

---------------------------------------------------------------------------------------------------
Version: 0.1.1
Date: 13. 12. 2017
  Changes:
    - Update for Factorio 0.16
    - Temporarily remove support for QuickItemSwap to work around [a bug in the current Factorio experimental](https://forums.factorio.com/viewtopic.php?f=182&t=54567&p=321491).  To counteract this, support is now provided in QuickItemSwap itself.

---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 11. 12. 2017
  Changes:
    - First release.


control.lua
config = require "config"

local function qis_register_events()
    if not remote.interfaces['QuickItemSwap'] then
        return
    end
    local events = remote.call('QuickItemSwap', 'get_events')

    script.on_event(events.on_qis_mappings_reset, qis_apply_patch)
end

local function qis_apply_patch()
    if not remote.interfaces['QuickItemSwap'] then
        return
    end

    return remote.call("QuickItemSwap", "apply_patch", {
        categories = {
            trains = {
                groups = {
                    trains = {
                        default_order = 100,
                        items = {
                            [config.PREFIX .. "electric-locomotive"] = { order = 150 }
                        }
                    }
                }
            }
        },
    }, "BatteriesNotIncluded")
end

--local function on_init()
--    qis_register_events()
--    qis_apply_patch()
--end
--
--local function on_load()
--    qis_register_events()
--end

script.on_init(qis_register_events)
script.on_load(qis_register_events)
--script.on_configuration_changed(qis_apply_patch)

-- Register events.
local function has_recoverable_battery(entity)
    local burner  = entity.burner
    return (
        burner and -- Has a burner
        burner.currently_burning and -- Is currently burning fuel
        burner.currently_burning.fuel_category == 'stored-electricity' and -- Is our fuel type
        burner.currently_burning.burnt_result  -- Has a burnt result?
    )
end

local function recover_battery(event)
    if has_recoverable_battery(event.entity) then
        event.buffer.insert({name=event.entity.burner.currently_burning.burnt_result.name, count=1})
    end
end

local function pre_recover_battery(event)
    if has_recoverable_battery(event.entity) then
        event.entity.burner.burnt_result_inventory.insert({name=event.entity.burner.currently_burning.burnt_result.name, count=1})
        event.entity.burner.currently_burning = nil
    end
end



script.on_event(defines.events.on_player_mined_entity, recover_battery)
script.on_event(defines.events.on_robot_mined_entity, recover_battery)


prototypes/vehicles.lua
require "util"

--local config = require "../config"

local locomotive = table.deepcopy(data.raw.locomotive.locomotive)
locomotive.name = config.PREFIX .. "electric-locomotive"
locomotive.reversing_power_modifier = 1.0 -- Electric engines don't care
locomotive.burner.fuel_category = "stored-electricity"
locomotive.burner.fuel_inventory_size = 2
locomotive.burner.burnt_inventory_size = 2
locomotive.burner.smoke = nil
locomotive.color = { r = 0.333, g = 0.733, b = 1.00, a = 0.5 }
locomotive.minable.result = locomotive.name
locomotive.working_sound = {
    sound = {
        filename = config.BASEDIR .. "/sounds/electric-locomotive.ogg",
        volume = 0.4,
    },
    match_speed_to_activity = true,
}
data:extend({
    locomotive,

    {
        type = "item-with-entity-data",
        name = locomotive.name,
        icon = config.BASEDIR .. "/graphics/icons/electric-locomotive.png",
        icon_size = 32,
        subgroup = "transport",
        order = "a[train-system]-f[electric-locomotive]",
        place_result = config.PREFIX .. "electric-locomotive",
        stack_size = 5,
    },

    {
        type = "recipe",
        icon = config.BASEDIR .. "/graphics/icons/electric-locomotive.png",
        icon_size = 32,
        name = locomotive.name,
        energy = 10,
        ingredients = {
            { "electric-engine-unit", 20 },
            { "electronic-circuit", 10 },
            { "steel-plate", 30 },
            { "accumulator", 1 },
        },
        result = locomotive.name,
        enabled = false,
    },
})


prototypes/batteries.lua
--local config = require "config"
local make_battery_pack = function(args)
        -- Time for charging is assuming a base charging station (speed=1) of 500 kw
        -- So, 30 MJ / 500 kW = 60 seconds.

    local name = args.name
    local prefixed = config.PREFIX .. name
    local mj = args.mj
    local ingredients = args.ingredients

    args.name = nil
    args.mj = nil
    args.ingredients = nil

    local empty = {
        type = "item",
        name = prefixed .. "-empty",
        icon = config.BASEDIR .. "/graphics/icons/" .. name .. "-empty.png",
        icon_size = 32,
        subgroup = "intermediate-product",
        order = "j[battery]-a[" .. name .. "-empty]",
        stack_size = args.stack_size or 10,
    }

    local full = {
        type = "item",
        name = prefixed,
        icon = config.BASEDIR .. "/graphics/icons/" .. name .. ".png",
        icon_size = 32,
        fuel_category = "stored-electricity",
        subgroup = "intermediate-product",
        order = "j[battery]-a[" .. name .. "]",
        stack_size = empty.stack_size,
        burnt_result = empty.name,
        fuel_value = mj .. "MJ",
        fuel_acceleration_multiplier = 1.0,
        fuel_top_speed_multiplier = 1.0,
    }

    for k, v in pairs(args) do
        full[k] = v
    end

    local recharge = {
        type = "recipe",
        name = full.name .. "-recharge",
        enabled = false,
        energy_required = mj * config.CHARGE_TIME_MULTIPLIER,
        ingredients = { { empty.name, 1 } },
        result = full.name,
        category = "electric-charging",
        icon_size = 32,
    }

    return empty, full, recharge
end

local empty, full, recharge
empty, full, recharge = make_battery_pack {
    name = "battery-pack",
    mj = 20,
    fuel_acceleration_multiplier = 1.6,
    fuel_top_speed_multiplier = 1.1,
}

data:extend{
    empty, full, recharge,
    {
        type = "recipe",
        name = empty.name,
        enabled = false,
        result = empty.name,
        category = "crafting",
        energy_required = 10,
        ingredients = {
            { "battery", 5 },
            { "plastic-bar", 2 },
            { "advanced-circuit", 2 },
        },
        icon_size = 32,
    }
}

empty, full, recharge = make_battery_pack{
    name = "battery-pack-2",
    mj = 75,
    fuel_acceleration_multiplier = 2.0,
    fuel_top_speed_multiplier = 1.25,
}

data:extend{
    empty, full, recharge,
    {
        type = "recipe",
        name = full.name,
        enabled = false,
        result = full.name,
        category = "crafting",
        energy_required = 10,
        ingredients = {
            { config.PREFIX .. "battery-pack", 5 },
            { "processing-unit", 5 },
            { "advanced-circuit", 5 },
        },
        icon_size = 32,
    }
}


prototypes/technology.lua
--local config = require "config"

local green_science = { { "automation-science-pack", 1 }, { "logistic-science-pack", 1 }, }
local blue_science = { { "automation-science-pack", 1 }, { "logistic-science-pack", 1 }, { "utility-science-pack", 1 }}

data:extend({
    {
        type = "technology",
        name = config.PREFIX .. "battery-pack",
        icon = config.BASEDIR .. "/graphics/icons/battery-pack.png",
        icon_size = 32,
        effects = {
            { type = "unlock-recipe", recipe = config.PREFIX .. "battery-pack-empty", },
            { type = "unlock-recipe", recipe = config.PREFIX .. "battery-pack-recharge", },
            { type = "unlock-recipe", recipe = config.PREFIX .. "charging-station", },
        },
        prerequisites = { "electric-energy-accumulators", "advanced-electronics", "electric-energy-distribution-1" },
        unit = {
            count = 200,
            ingredients = green_science,
            time = 20,
        },
        order = "e",
    },
    {
        type = "technology",
        name = config.PREFIX .. "electric-locomotive",
        icon = config.BASEDIR .. "/graphics/icons/electric-locomotive.png",
        icon_size = 32,
        effects = {
            { type = "unlock-recipe", recipe = config.PREFIX .. "electric-locomotive", },
        },
        prerequisites = { config.PREFIX .. "battery-pack", "railway" },
        unit = {
            count = 200,
            ingredients = green_science,
            time = 20,
        },
        order = "e-a",
    },
    {
        type = "technology",
        name = config.PREFIX .. "high-performance-battery-pack",
        icon = config.BASEDIR .. "/graphics/icons/battery-pack-2.png",
        icon_size = 32,
        effects = {
            { type = "unlock-recipe", recipe = config.PREFIX .. "battery-pack-2", },
            { type = "unlock-recipe", recipe = config.PREFIX .. "battery-pack-2-recharge", },
        },
        prerequisites = { config.PREFIX .. "battery-pack", "advanced-electronics-2", "electric-energy-distribution-2" },
        unit = {
            count = 300,
            ingredients = blue_science,
            time = 30,
        },
        order = "e-b",
    },
    {
        type = "technology",
        name = config.PREFIX .. "rapid-charging",
        icon = config.BASEDIR .. "/graphics/icons/rapid-charging-station.png",
        icon_size = 32,
        effects = {
            { type = "unlock-recipe", recipe = config.PREFIX .. "rapid-charging-station", },
        },
        prerequisites = { config.PREFIX .. "battery-pack", "advanced-electronics-2" },
        unit = {
            count = 300,
            ingredients = blue_science,
            time = 40,
        },
        order = "e-c",
    },

})


prototypes/chargers.lua
--local config = require "config"

local function make_charging_station(args)
    local name = args.name
    local prefixed = config.PREFIX .. name
    local crafting_speed = args.crafting_speed or 1
    local efficiency = args.efficiency or 1.0
    local energy_usage = args.energy_usage or (((config.BASE_CHARGING_KW * crafting_speed) / efficiency) .. "kW")
    local ingredients = args.ingredients
    local drain = args.drain

    args.speed = nil
    args.efficiency = nil
    args.energy_usage = nil
    args.ingredients = nil
    args.drain = nil

    local entity = {
        type = "furnace",
        name = prefixed,
        icon = config.BASEDIR .. "/graphics/icons/" .. name .. ".png",
        icon_size = 32,
        flags = { "placeable-neutral", "placeable-player", "player-creation" },
        minable = { mining_time = 1, result = prefixed },
        max_health = 150,
        corpse = "medium-remnants",
        collision_box = { { -0.9, -0.9 }, { 0.9, 0.9 } },
        selection_box = { { -1, -1 }, { 1, 1 } },
        dying_explosion = "medium-explosion",
        -- light = { intensity = 0.3, size = 7 },
        --        module_specification =
        --    {
        --      module_slots = 2,
        --      module_info_icon_shift = {0, 0.8}
        --    },
        --    allowed_effects = {"consumption", "speed", "productivity", "pollution"},
        crafting_categories = { "electric-charging" },
        result_inventory_size = 1,
        crafting_speed = crafting_speed,
        energy_usage = energy_usage,
        source_inventory_size = 1,
        energy_source = {
            type = "electric",
            usage_priority = "secondary-input",
            drain = drain,
            -- emissions = 0.005   -- FIXME: What should this be?
        },
        vehicle_impact_sound = { filename = "__base__/sound/car-metal-impact.ogg", volume = 0.65 },
        working_sound = {
            sound = {
                filename = "__base__/sound/accumulator-working.ogg",
                volume = 1
            },
            idle_sound = {
                filename = "__base__/sound/accumulator-idle.ogg",
                volume = 0.4
            },
            max_sounds_per_type = 5
        },
        working_visualisations =
        {
            {
                animation = {
                    -- May need layers = {...}
                    filename = config.BASEDIR .. "/graphics/entities/" .. name .. "-charge-animation.png",
                    width = 138,
                    height = 135,
                    line_length = 8,
                    frame_count = 24,
                    shift = { 0.46875, -0.640625 },
                    animation_speed = 0.5
                },
                light = { intensity = 0.3, size = 7, color = { r = 1.0, g = 1.0, b = 1.0 } },
            },
        },
        fast_replaceable_group = config.PREFIX .. "battery-charge-station",
        animation =
        {
            layers = {
                {
                    filename = config.BASEDIR .. "/graphics/entities/" .. name .. ".png",
                    priority = "high",
                    width = 124,
                    height = 103,
                    frame_count = 1,
                    shift = { 0.6875, -0.203125 }
                },
            }
        }
    }

    for k, v in pairs(args) do
        entity.k = v
    end


    local item = {
        type = "item",
        name = entity.name,
        icon = entity.icon,
        subgroup = "production-machine",
        order = "x[" .. name .. "]",
        place_result = entity.name,
        stack_size = 50,
        icon_size = 32,
    }

    local recipe = {
        type = "recipe",
        name = item.name,
        enabled = false,
        --        energy = 10,
        ingredients = ingredients,
        energy = energy,
        result = entity.name,
        category = "crafting",
        icon_size = 32,
        --        category = "advanced-crafting",
    }

    return { entity, item, recipe }
end


data:extend(make_charging_station {
    name = "charging-station",
    drain = "0kW",
    ingredients = {
        { "accumulator", 1 },
        { "copper-cable", 50 },
        { "electronic-circuit", 20 },
        { "advanced-circuit", 10 }
    }
})


data:extend(make_charging_station {
    name = "rapid-charging-station",
    drain = "0kW",
    crafting_speed = 2.5,
    efficiency = 0.8,
    ingredients = {
        { config.PREFIX .. "charging-station", 1 },
        { "advanced-circuit", 20, },
        { "processing-unit", 15 },
    }
})
